<launch>
  <arg name="valid_min_velocity" default="5.0"/>
  <arg name="detect_manual_engage" default="true"/>
  <arg name="time_delay_estimator_param" default="$(find-pkg-share time_delay_estimator)/config/general_time_delay_estimator_param.yaml"/>

  <!-- calibration adapter -->
  <include file="$(find-pkg-share calibration_adapter)/launch/calibration_adapter.launch.xml"/>

  <node pkg="time_delay_estimator" exec="general_time_delay_estimator" name="accel_time_delay_estimator" output="screen" respawn="true">
    <!-- input and response -->
    <param from="$(var time_delay_estimator_param)"/>
    <param name="data_name" value="accel"/>
    <param name="detect_manual_engage" value="$(var detect_manual_engage)"/>
    <param name="min_valid_value" value="0.05"/>
    <param name="max_valid_value" value="1.00"/>
    <param name="offset_value" value="0.0"/>
    <param name="accel/min_stddev_threshold" value="0.005"/>
    <remap from="~/input/control_mode" to="/vehicle/status/control_mode"/>
    <remap from="~/input/is_engage" to="/calibration/vehicle/is_engage"/>
    <remap from="~/input/input_cmd" to="/calibration/vehicle/acceleration_cmd"/>
    <remap from="~/input/input_status" to="/calibration/vehicle/acceleration_status"/>
    <remap from="~/output/time_delay" to="/vehicle/status/accel_time_delay"/>
  </node>

  <node pkg="time_delay_estimator" exec="general_time_delay_estimator" name="brake_time_delay_estimator" output="screen" respawn="true">
    <!-- input and response -->
    <param from="$(var time_delay_estimator_param)"/>
    <param name="data_name" value="brake"/>
    <param name="detect_manual_engage" value="$(var detect_manual_engage)"/>
    <param name="min_valid_value" value="0.05"/>
    <param name="max_valid_value" value="1.00"/>
    <param name="offset_value" value="0.0"/>
    <param name="brake/min_stddev_threshold" value="0.005"/>
    <remap from="~/input/control_mode" to="/vehicle/status/control_mode"/>
    <remap from="~/input/is_engage" to="/calibration/vehicle/is_engage"/>
    <remap from="~/input/input_cmd" to="/calibration/vehicle/brake_cmd"/>
    <remap from="~/input/input_status" to="/calibration/vehicle/brake_status"/>
    <remap from="~/output/time_delay" to="/vehicle/status/brake_time_delay"/>
  </node>

  <node pkg="time_delay_estimator" exec="general_time_delay_estimator" name="steer_time_delay_estimator" output="screen" respawn="true">
    <!-- input and response -->
    <param from="$(var time_delay_estimator_param)"/>
    <param name="data_name" value="steer"/>
    <param name="detect_manual_engage" value="$(var detect_manual_engage)"/>
    <param name="min_valid_value" value="0.05"/>
    <param name="max_valid_value" value="1.00"/>
    <param name="offset_value" value="0.0"/>
    <param name="steer/min_stddev_threshold" value="0.0025"/>
    <remap from="~/input/control_mode" to="/vehicle/status/control_mode"/>
    <remap from="~/input/is_engage" to="/calibration/vehicle/is_engage"/>
    <remap from="~/input/input_cmd" to="/calibration/vehicle/steering_angle_cmd"/>
    <remap from="~/input/input_status" to="/calibration/vehicle/steering_angle_status"/>
    <remap from="~/output/time_delay" to="/vehicle/status/steer_time_delay"/>
  </node>
  <node pkg="time_delay_estimator" exec="correlation_visualizer.py" name="correlation_visualizer" output="screen"/>
</launch>
